### PR TITLE
N Queen Problem Code (C++)

### PR DESCRIPTION
**Pull request by AasthaSinha2305**
@tarun26091999 @rudrakshi99 @anubhavitis 
I have added the code for n queen problem in backtracking section

- added the code in Code -> C++ folder.
- updated the backtracking readme
- commented each and every line properly
- 2 test cases added

issue #381

### CODE REVIEW

File: Code/C++/n_queen_problem.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
+/*
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:2]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ N queen problem is a standard backtracking problem where according to the
+ users input we have to place N queen of a (N X N) board so that they do not attack each other.
+
+ Movement of a queen on any board:  Diagonally or horizontally or vertically
+
+ Backtracking approach: In this method we place one queen row by row using recursion, Before placing the queen we check
+                        whether it will be attacked by other queen or not,
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:9]: press a tab(4 spaces to provide spaces here)

File: Code/C++/n_queen_problem.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+using namespace std;
+//A queen is safe if there is not queen horizontally, verically and diagonal on both sides of eleements
+bool isSafe(int chessboard[][100], int row, int col, int n){
+    //storing the position in x and y; 
+    int x = row;
+    int y = col;
+
+    //checking left diogonal
+    while(x>=0 && y>=0){
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:30]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        }
+        //cordinate we can get of row by reducing 1 and of column by reducing 1;
+        x--;
+        y--;
+    }
+
+    //Again storing the position in x and y
+    x = row;
+    y = col;
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:43]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                return true;
+            }
+            // if next queen connot be placed in any column, then we remove the queen we have placed earlier and try placing it in different column of this row
+            chessboard[row][j]=0;
+        }
+    }
+    // if control comes here that mean we are not able to place all the queen 
+    // so return false
+    return false;
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:117]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+int main(){
+    //define n
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:121]: Remove unnecessary comments

File: Code/C++/n_queen_problem.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    //function defined for backtracking approach
+    recursiveBacktracking(chessboard,0,n);
+    return 0;
+}
+
+
+
+
+/*
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:138]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+INPUT: 4
+
+OUTPUT: 
+
+* Q * *
+* * * Q
+Q * * *
+* * Q *
+
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:150]: Remove extra line

File: Code/C++/n_queen_problem.cpp [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+* Q * * *
+
+
+*/
+
+
+/*
+
+TIME COMPLEXITY: T(n) = O(N!)
+
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:239]: Add space complexity.

File: Code/C++/n_queen_problem.cpp [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+* * Q * *
+Q * * * *
+* * * Q *
+* Q * * *
+
+
+*/
+
+
+/*
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:236]: Include time and space complexity in same comment

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Comment by AasthaSinha2305** [comment]
@raksha009 I did the requested changes, please verify.

File: Code/C++/n_queen_problem.cpp [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        }
+        //cordinate we can get of row by reducing 1 and of column by reducing 1;
+        x--;
+        y--;
+    }
+    //Again storing the position in x and y for reference
+    x = row;
+    y = col;
+    //checking right diagonal
+    while(x>=0 && y<n){
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:38]: Add spaces around the operators like =, > , <

File: Code/C++/n_queen_problem.cpp [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        if(chessboard[x][y]==1){
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:39]: Add spaces around the operators like ==

File: Code/C++/n_queen_problem.cpp [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    //defining a 2d matrix 100 X 100 (Can be defined as per your input range)
+    int chessboard[100][100];
+    //initializing the 2d matrix with initial value 0
+    memset(chessboard,0,sizeof(chessboard));
+    //function defined for backtracking approach
+    recursiveBacktracking(chessboard,0,n);
+    return 0;
+}
+/*
+    TEST CASE 1:
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:121]: Plz align text along the horizontal.

**Review by raksha009 — CHANGES_REQUESTED** [review]

File: Code/C++/n_queen_problem.cpp [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+    whether it will be attacked by other queens or not,
+    if not, we can place the queen in that particular coulumn of that row
+    if the queen is being at that partiular column, we try placing it in the next column of that row,
+    if it is not possible to place the queen in any column of that row, we backtrack and change the position of the queen that was placed in the previous row and so on.
+    in the similar way we place all the N queens on the (N X N) board. 
+*/
+#include <iostream>
+#include <cstring>
+using namespace std;
+//A queen is safe if there is not queen horizontally, verically and diagonal on both sides of eleements
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:18]: Add a line after line 17

File: Code/C++/n_queen_problem.cpp [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+* * Q * *
+
+
+* * * * Q
+* * Q * *
+Q * * * *
+* * * Q *
+* Q * * *
+
+TIME COMPLEXITY: O(N!)
```
Comment (raksha009) [inline Code/C++/n_queen_problem.cpp:208]: Specify what is n

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Comment by AasthaSinha2305** [comment]
@raksha009 please verify this

**Review by rudrakshi99 — APPROVED** [review]

**Review by raksha009 — APPROVED** [review]
LGTM!

**Pull request merged**